torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,795 @@
+        aten.ones_like.default,
+        aten.t.default,
+        aten.view.default,
+        aten.reshape.default,
+        aten.detach.default,
+        aten.clone.default,
+    ]
+
+    dp_strategy_map = {}
+    batch_dim_analzer = BatchDimAnalyzer(batch_dim)
+    placeholder_idx = 0
+    num_param_grad = 0
+
+    # first we backward propagate to mark the param gradients sharding
+    # with tag_grad node helps and then delete the tag_grad nodes
+    for node in reversed(list(train_step_graph.graph.nodes)):
+        # find a param_grad node via the tagging
+        if node.target == torch.ops._spmd.tag_grad.default:
+            cur_node = node
+            while cur_node.target in non_compute_ops:
+                cur_node = cur_node.args[0]
+                partial_strategy = _gen_partial_strategy(mesh)
+                dp_strategy_map[cur_node] = DataParallelStrategy(
+                    NodeType.GRAD, [partial_strategy]
+                )
+            num_param_grad += 1
+            # remove the tag_grad node from graph
+            node.replace_all_uses_with(node.args[0])
+            train_step_graph.graph.erase_node(node)
+
+            if num_param_grad == num_params:
+                # early break if we have already processed all param_grads
+                break
+
+    train_step_graph.recompile()
+
+    # next we forward propagate to mark all the sharding
+    for node in train_step_graph.graph.nodes:
+        if node.op == "placeholder":
+            if "val" not in node.meta:
+                # NOTE: There're certain cases where the placeholder nodes do
+                # not have real tensor values:
+                # 1. optimizer states can be None sometimes, i.e. SGD with
+                #    no momentum, optimizer states populate `momentum` state
+                #    as None, the full graph we get from `compile` would have
+                #    None as the placeholder value
+                # 2. function args might not only contain params or activations,
+                #    but also contain other non-tensor inputs, i.e. the model
+                #    and optimizer instances baked in as a placeholder, there might
+                #    also be some scalar argument which is not a tensor
+                #
+                # For the above cases, we create a NON_TENSOR stratgy so that we
+                # know it's not a tensor and we don't need to shard it
+                dp_strategy_map[node] = DataParallelStrategy(NodeType.NON_TENSOR, [])
+
+            elif placeholder_idx < num_params:
+                # during compilation there's an assumption that the first num_params
+                # placeholders should be parameters
+                shard_strategy = _gen_shard_strategy(mesh, 0)
+                replica_strategy = _gen_replicate_strategy(mesh)
+                dp_strategy_map[node] = DataParallelStrategy(
+                    NodeType.PARAM, [replica_strategy, shard_strategy]
```

Comment:
noob question: what exactly does this mean? Specifically, the list of strategies passed to DataParallelStrategy. Is this saying that the 0th dim is replicated and the 1st dim is sharded? (That doesn't seem right). Or is it saying that we should try to replicate it first, and if that doesn't work, shard it? (But that also doesn't seem right?)

Reply:
I should clarify more in the `DataParallelStrategy` comment, will add it. Basically a `PlacementStrategy` means one way to distribute the tensor, a StrategyList(DataParallelStrategy) that have a `List[PlacementStrategy]` means there could be multiple ways to distribute the tensor. 

In this data parallel case, for model parameters, there're two possible ways to distribute the parameters: 1. replicate the parameter over a set of devices (DDP like behavior) 2. shard the parameter on its tensor dimension 0 over a set of devices (FSDP like behavior). 

So here we create a `DataParallelStrategy` that have two possible ways of sharding this parameter during `BuildDataParallelStrategy`. Then later in `mark_data_parallel_sharding`, according to whether the DataParallelStyle is replicate or fully_shard, we assign only one `PlacementStrategy` (replicate use the `replica_strategy`, fully_shard use the `shard_strategy`) to the node and use this strategy to do the real sharding.